db/migrate/20200217235235_populate_application_authentications.rb

```diff
@@ -0,0 +1,41 @@
+class PopulateApplicationAuthentications < ActiveRecord::Migration[5.2]
+  class ApplicationType < ActiveRecord::Base
+  end
+  class Authentication < ActiveRecord::Base
+    belongs_to :application_type
+    belongs_to :source
+  end
+  class Endpoint < ActiveRecord::Base
+    belongs_to :source
+  end
+  class Source < ActiveRecord::Base
+    belongs_to :source_type
+    has_many :applications
+  end
+  class SourceType < ActiveRecord::Base
+  end
+
+  def up
+    Authentication.all.each do |auth|
+      next if auth.resource_type != "Endpoint"
+
+      endpoint = Endpoint.find(auth.resource_id)
+      next if endpoint.nil?
+
+      source_type_name = endpoint.source&.source_type&.name
+
+      applications = endpoint.source&.applications || []
+      applications.each do |application|
+        next if application.tenant&.id.nil?
```

Comment:
Do we need a `Tenant` stub here?

Reply:
`Application` needs a `belongs_to :tenant` or we can use `application.tenant_id` to avoid instantiating a Tenant object